restapi/src/main/java/io/stargate/web/docsapi/service/query/FilterExpression.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.stargate.web.docsapi.service.query;
+
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.util.ExprFactory;
+import io.stargate.web.docsapi.service.query.condition.BaseCondition;
+import java.util.*;
+import java.util.function.Function;
+import org.immutables.value.Value;
+
+/** Filter expression that stands as the node in the Expression tree. */
+@Value.Immutable
+public abstract class FilterExpression extends Expression<FilterExpression> {
```

Comment:
Conceptually the type parameter of `Expression` appears to be the type of expression variable _names_, I think. Cf. `Variable.replaceVars`.

So, in our case it can probably be `Void` since we do not use variables.

Alternatively, since filters apply "conditions" to document properties, this type parameter could reasonably be `FilterPath`, I guess. The idea being that we should be able to resolve an `Expression<FilterPath>` if we have known values for all paths referenced by it (and its child expressions).

WDYT?

Reply:
I would agree here. I was also mentioning this in the talk we had. Let's fix it to `Void` for now, and then let's see along the path if this would change. OK?

Reply:
OK decided not to do it at the end. Then I can not collect all the filter expression from the tree and this was a very neat feature of having it like this.. Let's figure it out along the way..

Reply:
Ah, so you use the `collectK` method for gathering all the `FilterExpression` objects? Yes, that's a nice approach. Sorry, I missed that.